tools/generate-webview-js

```diff
@@ -30,5 +44,117 @@ export default \`
+    /* file: outputFilename, */
+  });
+
+  // In general, Babel may return multiple output chunks. This particular
+  // compilation setup, however, should return only a single chunk.
+  assert.strictEqual(output.length, 1);
+
+  // Extract and wrap that chunk's code.
+  const wrappedOutput = wrapCode(output[0].code);
+
+  // Write the transformed code.
+  await writeFileAsync(outputFilename, wrappedOutput);
+})().catch(
+  // eslint-disable-next-line no-console
+  console.error,
+);
+
+/* TODO:
+
+  At present, `webview/script.js` wraps this script's output; in turn this
+  script applies a wrapper around (the compiled text of) `webview/js.js`.
 
-fs.writeFileSync(outputFilename, output);
+  This is at least one layer of abstraction more than we need.
+
+  In the future, we'll probably want to make `js.js` compile to a single static
+  asset file. Only a <script> tag and a one-to-three-line invocation should need
+  to be emitted into the dynamically-rendered HTML.
+*/
+
+/**
+ * # Babel, core-js and polyfills
+ *
+ * The `@babel/preset-env` package relies on the definitions and compatibility
+ * declarations in `core-js` to select polyfills for rollup to include.
+ *
+ * Alas, this induces problems.
+ *
+ * The `core-js` package is dedicated to 100% compatibility, even on very old
+ * (ES3) engines. This means that its compatibility profile includes polyfills
+ * for marginal (read: irrelevant) functionality like `@@species` support
+ * [c#644].
+ *
+ * Furthermore, the compatibility table checks aren't run recursively on the
+ * dependencies of those polyfills; instead, the polyfills' dependencies are
+ * polyfilled automatically. [t] `core-js`'s feature-detection _should_ prevent
+ * these unnecessary polyfills from actually attaching to global objects; but
+ * Rollup is entirely oblivious to this, and will (as directed by `core-js`)
+ * cheerfully push into the final bundle all sorts of unnecessary-but-
+ * `require()`d polyfills containing all sorts of support functionality that,
+ * due to their runtime checks, can't be tree-shaken away. [c#388]
+ *
+ * It is possible to exclude individual polyfills on an item-by-item basis
+ * [b.excl]; but this will only block detection at the top level by
+ * @babel/preset-env. Even `core-js`'s own exclusion system [c.cloa] will not
+ * honor internal polyfill block requests. [c#388]
+ *
+ * It's bad enough that `core-js` inflates the bundle with things like IE8- bug
+ * workarounds [t] [c#388] and unstripped comments [t], and that it replaces
+ * fast, sufficient native code with complex, less-JITtable code. Worse, though,
+ * is that its aggressive replacement of native functionality can introduce
+ * slowdown even on formally-unrelated code paths. [b#5771] [c#306]
+ *
+ * For all that, `core-js` _does_ work. If we have another case like f65b50c85,
+ * where post-KitKat functionality is found to be in use, we may need to
+ * consider biting the bullet and shipping the bloat.
```

Comment:
I think this last bit isn't what we would actually do, for two reasons:
* Android 4.4 KitKat is down to 0.6% of our Android users, [as of last month](https://chat.zulip.org/#narrow/stream/48-mobile/topic/platform.20versions/near/786475). So if it came to a choice between some material degradation in either the developer experience, or the user experience on other versions, and simply dropping KitKat altogether... we'd drop it.
* Plus, in the case of webview browser features, the situation is much happier than that! Even among users on our oldest Android versions, the large majority of them have a WebView implementation which is a completely up-to-date Chrome.

If we had unbounded resources, we could always go for the third leg of the choice, namely to invest whatever engineering work was needed to make things good on the oldest KitKat browser without affecting anything else... but we have plenty of other things ahead of us to build and to fix that are important to much larger swathes of our users.

This is meant to be captured in the discussion at the top of `js.js`. Are there points it'd be useful to add there?

Reply:
The latter sentence of your first bullet point, maybe. Possibly also maybe a one-line reference along the lines of (*handwave*) "Automatic polyfills are an option, but one we've decided against for now; see `tools/build-webview` for more details."

I'm happy to strike the quoted paragraph in favor of a pointer to `js.js`.

Reply:
I went to go make that followup change... and ended up writing rather more than that. :slightly_smiling_face: Sent as #3666 .